skbio/alignment/_path.py

```diff
@@ -0,0 +1,410 @@
+            test_obj = PairAlignPath.from_bits(np.isin(byte_arr, gap_chars))
+            if (
+                self.lengths.shape != test_obj.lengths.shape
+                or (self.lengths != test_obj.lengths).any()
+            ):
+                raise ValueError(
+                    "Provided sequences do not match existing segment lengths."
+                )
+            elif (
+                self.states.shape != test_obj.states.shape
+                or (self.states != test_obj.states).any()
+            ):
+                raise ValueError(
+                    "Provided sequences do not match existing states. Consider the "
+                    "order of the provided sequences."
+                )
+
+        cigar = ""
+        lengths = self.lengths
+        gaps = np.squeeze(self.states)
+        codes = ["M", "I", "D", "P"]
+        if seqs is not None:
+            query = str(seqs[0])
+            ref = str(seqs[1])
+            for qchar, rchar in zip(query, ref):
+                if qchar == "-" and rchar == "-":
+                    cigar += "P"
+                elif qchar == rchar:
+                    cigar += "="
+                elif qchar == "-":
+                    cigar += "I"
+                elif rchar == "-":
+                    cigar += "D"
+                else:
+                    cigar += "X"
+            return self._run_length_encode(cigar)
+        else:
+            for i, length in enumerate(lengths):
+                cigar += str(length) + codes[gaps[i]]
+            return cigar
+
+    @classonlymethod
+    def from_cigar(cls, cigar):
+        r"""Create a pairwise alignment path from a CIGAR string.
+
+        Parameters
+        ----------
+        cigar : str
+            CIGAR format string used to build the PairAlignPath.
+        """
+        # Make sure cigar is not empty.
+        if not cigar:
+            raise ValueError("CIGAR string must not be empty.")
+
+        lengths = []
+        gaps = []
+        current_length = 0
+        mapping = {
+            "M": 0,
+            "I": 1,
+            "D": 2,
+            "P": 3,
+            "=": 0,
+            "X": 0,
+            "N": 2,
+            "S": 1,
+            "H": 3,
+        }
+        no_ones = True
+        for char in cigar:
+            if char.isdigit():
+                no_ones = False
+                current_length = current_length * 10 + int(char)
+            elif char in mapping:
+                if no_ones:
+                    lengths.append(current_length + 1)
+                else:
+                    lengths.append(current_length)
+                gaps.append(mapping[char])
+                current_length = 0
+                no_ones = True
+            else:
+                raise ValueError("Invalid characters in CIGAR string.")
+        lengths, gaps = cls._fix_arrays(lengths=np.array(lengths), gaps=np.array(gaps))
+        return cls(lengths, gaps, [0, 0])
+
+    def _run_length_encode(self, string_in):
```

Comment:
This method does not use `self`. Therefore, it does not have to be a method of the class. Instead, it can be a standalone function.